### PR TITLE
[Web] Accept charset in content-type for json messages

### PR DESCRIPTION
**Pull request by ibizaman**
Trac: https://dev.deluge-torrent.org/ticket/3521

### CODE REVIEW

**Review by cas-- — CHANGES_REQUESTED** [review]
Thanks for reporting and suggesting a fix, I have provided some feedback to improve it

File: deluge/ui/web/json_api.py [hunk 1]
```diff
@@ -185,12 +185,40 @@ def _on_rpc_request_failed(self, reason, response, request):
...
+
+        return (splitted[0], splitted[1])
+
     def _on_json_request(self, request):
         """
         Handler to take the json data as a string and pass it on to the
         _handle_request method for further processing.
         """
-        content_type = request.getHeader(b'content-type').decode()
+        content_type, charset = self._parse_content_type(request.getHeader(b'content-type').decode())
```
[the remark this answers is not in the document]
Comment (ibizaman) [reply deluge/ui/web/json_api.py:221]: Good call. I agree with your proposal although TBH I don’t know which one to choose.

Being more strict and asserting the charset to be utf-8 would probably avoid some headaches down the line by failing early.

But on the other side being more lenient can avoid badly configured clients to be blocked.

My gut tells me being stricter is better but what do you think? Are there some similar choices that had been made on the past?

File: deluge/tests/test_webserver.py [hunk 2]
```diff
@@ -25,6 +25,13 @@
 class TestWebServer(WebServerTestBase, WebServerMockBase):
     @pytest_twisted.inlineCallbacks
     def test_get_torrent_info(self):
+        self._test_get_torrent_info('application/json')
+
+    @pytest_twisted.inlineCallbacks
+    def test_get_torrent_info_charset(self):
+        self._test_get_torrent_info('application/json;charset=utf-8')
+
+    def _test_get_torrent_info(self, content_type):
```
[the remark this answers is not in the document]
Comment (ibizaman) [reply deluge/tests/test_webserver.py:34]: Ah I totally missed that. I’ll change it.

File: deluge/ui/web/json_api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cas--) [reply deluge/ui/web/json_api.py:221]: It's safe to assume that all `application/json` content type will be `utf-8`, in fact the charset parameter is not part of the standard: https://github.com/request/request/issues/383#issuecomment-18455635

So let's keep it simple, you could still use `cgi.parse_header` and ignore the second tuple value using the `_`:

```
content_type, _ = cgi.parse_header(header-content-type)
```

[the remark this answers is not in the document]
Comment (ibizaman) [reply deluge/ui/web/json_api.py:221]: Oh I had no idea about that. Cool let’s keep it simple. I’ll definitely add a comment about it.

File: deluge/tests/test_webserver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ibizaman) [reply deluge/tests/test_webserver.py]: So instead, I wanted to check that the content-type was valid for `application/json` and `application/json;charset=utf-8` so I created a new test function for it. I'm happy to change that in any way, just tell me.

File: deluge/ui/web/json_api.py [hunk 3]
```diff
@@ -190,7 +191,9 @@ def _on_json_request(self, request):
         Handler to take the json data as a string and pass it on to the
         _handle_request method for further processing.
         """
-        content_type = request.getHeader(b'content-type').decode()
+        # We assume the all `application/json` content type will use
+        # the `utf-8` charset.
```
Comment (cas--) [inline deluge/ui/web/json_api.py:195]: I appreciate the addition of a comment but without any context about charset it is not worth mentioning here as comment. We should always err towards less comments unless the code we write is unclear but even then refactoring the code might help. 

The actual reasoning for making this change should be stored in the git commit message so we then use git blame in future if we are uncertain about this line of code. We also have the tests that could improved with a docstring to quickly outline reasoning.



I prefer that the test docstring mentions this and in the commit message provides more details about reasoning.

File: deluge/tests/test_json_api.py [hunk 4]
```diff
@@ -98,6 +98,17 @@ def test_on_json_request_invalid_content_type(self):
         with pytest.raises(JSONException):
             json._on_json_request(request)
 
+    def test_on_json_request_valid_content_type(self):
```
Comment (cas--) [inline deluge/tests/test_json_api.py:101]: I would split this into two tests, helps with understanding should a test fail
```
test_on_json_request_content_type
"""Ensure content-type application/json is accepted"""

test_on_json_request_content_type_charset
"""Ensure content-type parameters such as charset are ignored"""
```

**Review by cas-- — APPROVED** [review]
Thanks for patience with this, looks excellent :+1:

**Comment by ibizaman** [comment]
@cas--  With pleasure!

**Comment by ibizaman** [comment]
Thanks @cas--

**Pull request closed without merging**